Yuhsuan/fitting init val

YuHsuan-Hwang opened the pull request:
**Description**

* What is implemented or fixed? Mention the linked issue(s), if any.

  Supports initial value auto generation for image fitting.

* How does this PR solve the issue? Give a brief summary.

  * When there are NaN values in the request message `FittingRequest.initial_values`, `ImageFitter.SetInitialValues` returns false and initial values are regenerated using `ImageFitter.CalculateInitialValues`.
  * Added a new class `InitialValueCalculator` for generating initial values.
    * A KMeans++ clustering method is used to classify data points into different sources.
    * The method of moments is calculated twice to estimate Gaussian parameters. The first calculation obtains the source area and select a circular area so that the estimate will not be affect by the image boundaries. The second calculation obtains the Gaussian parameters based on the selected area.
    * The threshold is set to 4 * MAD by default. It is lowered when the result is not ideal, for example, when there are too few data points, KMeans++ produces fewer groups than needed, the method of moments obtains extremely large FWHM, or most part of a source is outside the image.
  * The result of the generated initial values are added in the response message `FittingResponse.log`.

* Are there any companion PRs (frontend, protobuf)?

  Frontend PR: https://github.com/CARTAvis/carta-frontend/pull/2546

ToDo: update documentation, update change log

**Checklist**

- [x] changelog updated / no changelog update needed
- [x] e2e test passing / corresponding fix added / new e2e test created
- [ ] ICD test passing / ~corresponding fix added~ / ~new ICD test created~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] ~protobuf version bumped~ / protobuf version not bumped
- [x] added reviewers and assignee
- [x] GitHub Project estimate added

YuHsuan-Hwang committed:
check init val inputs

YuHsuan-Hwang committed:
add new class; rename dir

YuHsuan-Hwang committed:
apply method of moments

YuHsuan-Hwang committed:
apply offset when fitting with fov or regions

YuHsuan-Hwang committed:
fix success status

YuHsuan-Hwang committed:
resolve conflict and merge dev

YuHsuan-Hwang committed:
apply k means plus plus to find centroid of each gaussian component

YuHsuan-Hwang committed:
consider negative amp; fix random

YuHsuan-Hwang committed:
apply threshold with mad

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val

YuHsuan-Hwang committed:
tmp set centroid center as init val center

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val

YuHsuan-Hwang committed:
fix image generator version

YuHsuan-Hwang committed:
fix doc

YuHsuan-Hwang committed:
fix ci

YuHsuan-Hwang committed:
set parameters to multi component

YuHsuan-Hwang committed:
support multi component method of moments

YuHsuan-Hwang committed:
support multi component initial value calculation

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val

YuHsuan-Hwang committed:
handle insufficient data point in kpp; adjust logging

YuHsuan-Hwang committed:
remove components when the generated params are not valid

YuHsuan-Hwang committed:
minor refactor

YuHsuan-Hwang committed:
adjust threshold when invalid params generated

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val

YuHsuan-Hwang committed:
fix naming style

confluence commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -188,6 +226,15 @@ void ImageFitter::SetInitialValues(
+
+bool ImageFitter::CalculateInitialValues(std::vector<CARTA::GaussianComponent>& initial_values) {
+    InitialValueCalculator* calculator =
+        new InitialValueCalculator(_fit_data.data, _fit_data.width, _fit_data.n / _fit_data.width, _fit_data.offset_x, _fit_data.offset_y);
```
I suggest passing a single FitData struct parameter into this constructor and unpacking it in the constructor.

These helper structs should also be moved from this file to a separate file in this directory, e.g. `ImageFitting/Util.[h|cc]` so that both the fitter and the calculator can use them.

YuHsuan-Hwang replied:
Changed the input parameter to FitData.

confluence commented on src/ImageFitting/InitialValueCalculator.cc:
```diff
@@ -0,0 +1,328 @@
+    _width = width;
+    _height = height;
+    _offset_x = offset_x;
+    _offset_y = offset_y;
```
Please move these into an initializer list (in addition to passing in one FitData parameter). This should also be done for the other constructors (and the structs themselves should have constructors), but that may be outside the scope of this PR.

YuHsuan-Hwang replied:
@confluence For `the other constructors`, do you suggest we apply initializer list to any fitting related constructors, or do you mean codebase not related to fitting?

YuHsuan-Hwang replied:
Applied initializer list to the calculator constructor.

confluence replied:
> @confluence For `the other constructors`, do you suggest we apply initializer list to any fitting related constructors, or do you mean codebase not related to fitting?

I meant the other constructors in these image fitting files -- since they aren't modified by this PR, it's not strictly related. I will re-review.

confluence commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -48,7 +48,35 @@ bool ImageFitter::FitImage(size_t width, size_t height, float* image, double bea
+        spdlog::info("Generating initial values for fitting.");
+        success = CalculateInitialValues(initial_values);
+        if (success) {
+            initial_value_log = InitialValueCalculator::GetLog(initial_values, _unit);
```
I would eliminate `ImageFitter::CalculateInitialValues` entirely, and construct the calculator object instance directly inside this function. This would also allow streamlining the design of the calculator, which is currently instantiated with some parameters, has one instance function run with additional parameters, and then has the instance discarded and has a static function run with the same initial parameters.

I suggest instead passing `initial_values` and `_image_std` into the calculator constructor as well, calling `CalculateInitialValues` here (no additional parameters), making `GetLog` an instance function, and calling it here with only `_unit`. Inside the calculator, the functions can use the initial values and standard deviation properties set on the instance by the constructor.

YuHsuan-Hwang replied:
Removed `ImageFitter::CalculateInitialValues`, passed `initial_values` and `_image_std` into the calculator constructor, and adjusted `GetLog` function.

confluence replied:
I've realised that half of this refactoring request was a bit silly; I have included an updated suggestion in my review.

confluence reviewed (changes requested):
I have some suggestions for the structure of the objects; other than that the code looks fine.

YuHsuan-Hwang committed:
move FitData to util

YuHsuan-Hwang committed:
pass single FitData struct parameter into the constructor

YuHsuan-Hwang committed:
apply initializer list to calculator constructor

YuHsuan-Hwang committed:
adjust doc

YuHsuan-Hwang committed:
pass initial values and std to calculator constructor

YuHsuan-Hwang committed:
remove fitter CalculateInitialValues function

YuHsuan-Hwang committed:
adjust calculator GetLog function

YuHsuan-Hwang committed:
add doc

YuHsuan-Hwang committed:
update changelog

YuHsuan-Hwang commented:
@confluence The structure is adjusted. Thanks for the suggestions.

confluence commented on src/ImageFitting/InitialValueCalculator.cc:
```diff
@@ -0,0 +1,329 @@
+      _height(fit_data->n / fit_data->width),
+      _offset_x(fit_data->offset_x),
+      _offset_y(fit_data->offset_y),
+      _initial_values(initial_values),
```
Sorry; I've realised that this is an in-out parameter, and setting it as a property is a bit awkward. I also think that the log construction can be merged into the public calculation function (since this is a log of the calculation process).

My revised suggestion:
* Don't pass `initial_values` to the constructor
* Do pass it into `CalculateInitialValues` as an in-out parameter
* Pass the image unit into the constructor instead
* Pass `initial_value_log` into `CalculateInitialValues` as another in-out parameter
* Build the log string inside `CalculateInitialValues`.

YuHsuan-Hwang replied:
I see your point. I'll adjust it.

YuHsuan-Hwang replied:
Revised based on the new suggestions.

confluence reviewed (changes requested):
I have an updated suggestion for the function signatures; I think that's the last thing.

pford commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -48,7 +48,36 @@ bool ImageFitter::FitImage(size_t width, size_t height, float* image, double bea
+    std::string initial_value_log = "";
+    if (!success) {
+        spdlog::info("Generating initial values for fitting.");
+        InitialValueCalculator* calculator = new InitialValueCalculator(&_fit_data, initial_values, _image_std);
```
Memory leak--`new` but no `delete`, use unique_ptr

confluence replied:
Good point; I missed this. But why is this a pointer at all? It could just be a normal object variable.

pford replied:
That's true, it does not get passed anywhere.

YuHsuan-Hwang replied:
I see. Changed to a normal object.

pford commented on src/ImageFitting/ImageFitter.h:
```diff
@@ -220,13 +201,13 @@ class ImageFitter {
      * @param offset_x X-axis offset from the fitting region to the entire image
      * @param offset_y Y-axis offset from the fitting region to the entire image
-     * @return A tuple of Gaussian parameters
+     * @return A tuple of Gaussian parameters: center x, center y, amplitude, FWHM x, FWHM y, and position angle
```
Add struct to Util.h instead of tuple, use in ImageFitter and InitialValueCalculator

YuHsuan-Hwang replied:
Added a new struct `GaussianParams` in `Util.h`. Also moved `GetGaussianComponent` to `Util.h` and applied it to the calculator.

YuHsuan-Hwang committed:
pass initial_value to CalculateInitialValues

YuHsuan-Hwang committed:
pass unit to contructor

YuHsuan-Hwang committed:
pass log to CalculateInitialValues

YuHsuan-Hwang committed:
call GetLog in CalculateInitialValues

YuHsuan-Hwang committed:
fix memory leak

YuHsuan-Hwang committed:
add struct GaussianParams and apply to fitter

YuHsuan-Hwang committed:
apply GaussianParams to calculator

YuHsuan-Hwang committed:
move GetGaussianComponent to util

YuHsuan-Hwang committed:
apply GetGaussianComponent to calculator

YuHsuan-Hwang committed:
update doc

YuHsuan-Hwang commented:
@confluence @pford Thanks for the comments. Please check my new changes, thanks!

confluence commented on src/ImageFitting/Util.h:
```diff
@@ -0,0 +1,82 @@
+    auto fwhm = Message::DoublePoint(params.fwhm_x, params.fwhm_y);
+    auto component = Message::GaussianComponent(center, params.amp, fwhm, params.pa);
+    return component;
+};
```
This should be an instance method of GaussianParams.

YuHsuan-Hwang replied:
Changed to method of GaussianParams.

confluence commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -492,8 +538,8 @@ void ImageFitter::ErrorHandler(const char* reason, const char* file, int line, i
 
-std::tuple<double, double, double, double, double, double> ImageFitter::GetGaussianParams(const gsl_vector* value_vector, size_t index,
-    std::vector<int>& fit_values_indexes, std::vector<double>& initial_values, size_t offset_x, size_t offset_y) {
+GaussianParams ImageFitter::GetGaussianParams(const gsl_vector* value_vector, size_t index, std::vector<int>& fit_values_indexes,
```
This should be an additional constructor of `GaussianParams`.

YuHsuan-Hwang replied:
@confluence I would prefer to remain this as a method in the fitter because it performs complex calculation related to fitting (checking `fit_values_indexes` and selecting fixed or fitted values). What do you think?

confluence replied:
OK, that's fair -- I was on the fence about this. I think it's fine to leave it where it is.

confluence commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -268,9 +307,11 @@ void ImageFitter::CalculateErrors() {
+        double amp = params.amp;
+        double fwhm_x = params.fwhm_x;
+        double fwhm_y = params.fwhm_y;
+        double pa = params.pa;
```
You can use a structured binding for this unpacking instead:
```
auto [center_x, center_y, amp, fwhm_x, fwhm_y, pa] = GetGaussianParams(...);
```

(But see also the previous comment about turning that function into a constructor.)

YuHsuan-Hwang replied:
Applied structured binding.

YuHsuan-Hwang committed:
move GetGaussianComponent to GaussianParams

YuHsuan-Hwang committed:
apply structured binding

YuHsuan-Hwang committed:
improve debug log

YuHsuan-Hwang committed:
fix incorrect guessed center for fov and regions

YuHsuan-Hwang commented:
@confluence @kswang1029 I have updated the PR:
* Adjusted codebase according to reviews.
* Fixed incorrect guessed center with FOV and regions.
* Improved debug log.

confluence commented on src/ImageFitting/ImageFitter.cc:
```diff
@@ -440,8 +478,13 @@ int ImageFitter::FuncF(const gsl_vector* fit_values, void* fit_data, gsl_vector*
 
-        double center_x, center_y, amp, fwhm_x, fwhm_y, pa;
-        std::tie(center_x, center_y, amp, fwhm_x, fwhm_y, pa) = GetGaussianParams(fit_values, k, d->fit_values_indexes, d->initial_values);
+        GaussianParams params = GetGaussianParams(fit_values, k, d->fit_values_indexes, d->initial_values);
+        double center_x = params.center_x;
+        double center_y = params.center_y;
+        double amp = params.amp;
+        double fwhm_x = params.fwhm_x;
+        double fwhm_y = params.fwhm_y;
+        double pa = params.pa;
```
The structured binding should also be used here.

YuHsuan-Hwang replied:
@confluence I get compile errors if applying here:
```shell
error: capturing a structured binding is not yet supported in OpenMP
  503 |                 double dx = i % d->width - center_x;
      |                                            ^
```
Or do you prefer we use `std::tie`?

confluence replied:
Oh, I see. Let's leave it alone for now.

confluence commented on src/ImageFitting/Util.h:
```diff
@@ -0,0 +1,81 @@
+     * @brief Create a Gaussian component sub-message.
+     * @return A Gaussian component sub-message
+     */
+    CARTA::GaussianComponent GetGaussianComponent() {
+        auto center = Message::DoublePoint(center_x, center_y);
+        auto fwhm = Message::DoublePoint(fwhm_x, fwhm_y);
+        auto component = Message::GaussianComponent(center, amp, fwhm, pa);
+        return component;
+    };
+};
```
I have a much cleaner suggestion for this. The only reason that this function exists is to group the center and fwhm parameters into sub-messages -- but these are all message internals, and a more appropriate place for them is `Util::Message`.

`Message::GaussianComponent` is *only* used by this code, so there's no reason for it to require the parameters in this format -- you can make that function take these six separate parameters instead, and do this packing inside.

Everywhere that this function is called, you already have access to the separate parameters, or you can use structured binding to get them instead of the single struct. So if you change the signature of `Message::GaussianComponent`, you can call it directly instead.

(Michaela is currently working on extending `Message` to functions which add submessages to messages, to simplify how this is done, but I think that level of refactoring is outside the scope of this PR. We can handle that in a later refactoring pass.)

YuHsuan-Hwang replied:
Ok. Let's work on this in future PRs.

kswang1029 reviewed (approved):
this looks good to me. e2e tests are adjusted accordingly.

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/fitting_init_val
